packages/@aws-cdk/custom-resource-handlers/lib/aws-synthetics-alpha/auto-delete-lambda-handler/index.ts

```diff
@@ -0,0 +1,77 @@
+/* eslint-disable import/no-extraneous-dependencies */
+import { LambdaClient, DeleteFunctionCommand } from '@aws-sdk/client-lambda';
+import { SyntheticsClient, GetCanaryCommand } from '@aws-sdk/client-synthetics';
+import { makeHandler } from '../../nodejs-entrypoint';
+
+const AUTO_DELETE_LAMBDA_TAG = 'aws-cdk:auto-delete-lambda';
+
+const lambda = new LambdaClient({});
+const synthetics = new SyntheticsClient({});
+
+export const handler = makeHandler(autoDeleteHandler);
+
+export async function autoDeleteHandler(event: AWSLambda.CloudFormationCustomResourceEvent) {
+  switch (event.RequestType) {
+    case 'Create':
+      break;
+    case 'Update':
+      return onUpdate(event);
+    case 'Delete':
+      return onDelete(event.ResourceProperties?.CanaryName);
+  }
+};
+
+async function onUpdate(event: AWSLambda.CloudFormationCustomResourceEvent) {
+  const updateEvent = event as AWSLambda.CloudFormationCustomResourceUpdateEvent;
+  const oldCanaryName = updateEvent.OldResourceProperties?.CanaryName;
+  const newCanaryName = updateEvent.ResourceProperties?.CanaryName;
+  const canaryNameHasChanged = (newCanaryName && oldCanaryName)
+    && (newCanaryName !== oldCanaryName);
+
+  // If the name of the canary has changed, CloudFormation will delete the canary
+  // and create a new one with the new name. So the old lambda will be isolated
+  // and we should try to delete the old lambdae. */
+  if (canaryNameHasChanged) {
+    return onDelete(oldCanaryName);
+  }
```

Comment:
I think a simpler implementation would be to return a different `PhysicalResourceId` (based off the CanaryName), then CFN will send us a `DELETE` event at the end of the UPDATE operation. (Is this current implementation rollbackable? Do all cleanup handlers suffer from this?)